CLUSTER.md

```diff
@@ -0,0 +1,490 @@
+the epoch. This new topology will then propagate to the various Failover Coordinators and nodes. There's no requirement for ordering in this
+flow -- either the target Failover Coordinator or the source Failover Coordinator can get the message first and the system will still achieve the desired end
+state.
+When an Failover Coordinator receives a heartbeat request from a node it does not own it returns a MOVED response to the correct Failover Coordinator
+(there will always be such an Failover Coordinator since the replacement is done in an atomic manner). As an implementation detail, after
+being moved between Failover Coordinators a node should wait double its normal time before initiating failover in order to avoid redundant
+fail-overs.
+If the target Failover Coordinator is still unaware of the new topology it might cause the shards to ping-pong a bit between the Failover Coordinators, so a node should contain logic to reduce
+the frequency of heartbeats after a moved response, but even if it doesn't this situation will self-resolve without
+causing any damage.
+
+Worth noting that this flow is the main reason why a Topology Director **must** ensure no concurrent updates to the same shard:
+- Shard X comprised of 3 nodes X1,X2,X3 belong to Failover Coordinator A
+- X is moved to Failover Coordinator B
+- All entities other than X3 get the updated topology
+- X1 sends a heartbeat as primary to Failover Coordinator **B**
+- X is moved back to Failover Coordinator A, but Failover Coordinator B is delayed in receiving the updated topology
+- X3 sends a heartbeat as primary to Failover Coordinator **A**
+While this situation will self-resolve when Failover Coordinator B receives the updated topology, Flotilla aims to guarantee a strongly 
+consistent topology - so it needs to prevent concurrent updates.
+
+![Move Shard](images/move_shards.jpg?raw=true "Move shard between Failover Coordinators")
+
+### Failover
+
+Every K milliseconds, every node sends a heartbeat to the Failover Coordinator which owns it. This heartbeat includes:
+- Replication offset (for replicas)
+- Last known cluster topology epoch
+- Last known Shard topology epoch
+- Role (primary/replica)
+
+The Failover Coordinator can respond with the following:
+- ACK: The node topology is up-to-date. The heartbeat data is saved in the Failover Coordinator along with the current tick.
+- NACK: The node Topology is out-of-date
+- MOVED: The node does not belong to the Failover Coordinator. This response includes the details of the relevant Failover Coordinator.
+
+The response includes:
+- The heartbeat results from all the nodes in the same shard (and their 'tick' value)
+- If needed, the updated cluster topology.
+
+If a replica node detects that its primary has not sent a heartbeat for more than N ticks it checks if it is the most
```

Comment:
in case the replica-primary keepalive is broken BUT the primary-FC heartbit is O.K - will this also be validated by the FC?
there are cases were the replication stream is broken (bug/route issues etc...) but that does not mean we should failover...

Reply:
The FC does not validate anything in the heartbeat beyond epoch correctness -- it's the responsibility of the individual nodes to trigger failovers.
In the case you outlined:
- If the replica <->  FC connectivity is ok the replica will see that the primary is still alive (since the FC shares that info in the heartbeat response) and not initiates failover
- If the replica <-> FC connectivity is down the replica can't promote itself even if it wanted to